webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/servers/RemoveNodeService.scala

```diff
@@ -327,69 +493,192 @@ class RemoveNodeServiceImpl(
     }
   }
 
+
   /**
-   * Uncertify node key if it was. Can be done once the node is deleted
+   * Look for the groups containing this node in their nodes list, and remove the node
+   * from the list
    */
-  private def removeKeyCertification(nodeId: NodeId): IOResult[LDIFChangeRecord] = {
+  def deleteFromGroups(nodeId: NodeId, modId: ModificationId, actor:EventActor): IOResult[Seq[ModifyNodeGroupDiff]]= {
+
     for {
-      con <- ldap
-      res <- con.modify(deletedDit.NODES.NODE.dn(nodeId.value), new Modification(ModificationType.REPLACE, LDAPConstants.A_KEY_STATUS, UndefinedKey.value))
+      _            <- NodeLoggerPure.Delete.debug(s"  - remove node ${nodeId.value} from his groups")
+      nodeGroupIds <- roNodeGroupRepository.findGroupWithAnyMember(Seq(nodeId))
+      deleted      <- ZIO.foreach(nodeGroupIds) { nodeGroupId =>
+                        for {
+                          nodeGroup    <- roNodeGroupRepository.getNodeGroup(nodeGroupId).map(_._1)
+                          updatedGroup =  nodeGroup.copy(serverList = nodeGroup.serverList - nodeId)
```

Comment:
what happens if a dynamic group update happens at the same time ? I don't see lock there

Reply:
I think that should be a function on the repo side "updateServerList" with 'add' and 'remove' parameter that uses the repo lock and update the server list in way that that is safe.

Reply:
I dunno if it should be done here, because we can also have the same problem with the UI, with an archive restoration (maybe not the same thing exactly, not the same impact)

Reply:
It was already like that, it's not a change from that PR, so it seems to work most of the time. 
I'm looking for implementing a "delete nodes" method in repos.

Reply:
Yes i think that if something should be done it should be in another PR because it goes beyond what we are going to achieve here

Reply:
Ah, too late, I added it :/

Reply:
lol ok !!

Reply:
Too bad it was merged in the first commit so i can't read the changes apart rereading the whole PR

Reply:
Hum, yeah, there was conflicting merge and I forgot to keep commits in the rebase